### ISSUE TITLE
Cannot read property 'setRowData' of null

### ISSUE THREAD

**Issue by ShankzPatil** [ag-grid/ag-grid#2030]
i am using ag grid with signalR implementation.
on first page load works fine but after user goes on second page and return to first page.
signalr notification received regarding data change. then i am updating rows at that time i am getting
Cannot read property 'setRowData' of null.

in debugging i found my gridoptions is containing api value as null 

![image](https://user-images.githubusercontent.com/22147418/32317130-c410b9a4-bfd8-11e7-9aac-0b2a7c41754c.png)

**Comment by jbarber2016** [comment]
You should be calling gridOptions.api.setRowData onGridRead. 

var gridOptions = {
  onGridReady: function() {
     this.gridOptions.api.setRowData(....)
  }
};

**Comment by ShankzPatil** [comment]
Now working fine after adding scope.watch on my gridData.

**Comment by damonus-prime** [comment]
Is there a suggested way for using the above onGridReady code in Angular?

I have this and it is never firing:

`setupGrid() {
        console.log('setting up grid');
        this.gridOptions = {
            columnDefs: this.columnDefs,
            headerHeight: 40,
            rowHeight: 40,
            onGridReady: () => {
                console.log('grid ready...');
                this.loadAll();
            }
        };
    }`

The first log statement fires, but the 'grid ready...' one does not.

**Comment by jbarber2016** [comment]
did you try this?  onGridReady: function(event) { console.log("the grid is now ready"); }

**Comment by damonus-prime** [comment]
@jbarber2016 just did.  No luck.  Also tried:

```
onGridReady(event) {
    console.log('grid ready');
}
```
with:
```
<ag-grid-angular
    (gridReady)="onGridReady($event);"
   ...
```

also no luck.  Tried with and without $event.

**Comment by jbarber2016** [comment]
what do you have in your packages.json?

**Comment by damonus-prime** [comment]
```
"dependencies": {
    "@angular/common": "4.2.6",
    "@angular/compiler": "4.2.6",
    "@angular/core": "4.2.6",
    "@angular/forms": "4.2.6",
    "@angular/http": "4.2.6",
    "@angular/platform-browser": "4.2.6",
    "@angular/platform-browser-dynamic": "4.2.6",
    "@angular/router": "4.2.6",
    "@ng-bootstrap/ng-bootstrap": "1.0.0-alpha.28",
    "ag-grid": "^13.3.1",
    "ag-grid-angular": "^13.3.0",
    "bootstrap": "4.0.0-alpha.6",
    "core-js": "2.4.1",
    "font-awesome": "4.7.0",
    "jquery": "3.2.1",
    "ng-jhipster": "0.2.7",
    "ng2-webstorage": "1.8.0",
    "ngx-cookie": "1.0.0",
    "ngx-infinite-scroll": "0.5.1",
    "reflect-metadata": "0.1.10",
    "rxjs": "5.4.2",
    "swagger-ui": "2.2.10",
    "tether": "1.4.0",
    "zone.js": "0.8.13"
  },
```

I should also mention I can get the grid working fine if I just use a plain `<div>`, create a Grid object in TypeScript and bind to the div.  The reason I am trying to switch is that I need to use an Angular component as a cell renderer, and if I try to do that with a plain div it's all "you are not using the component version of the grid, so you can't use a component as a CellRenderer" or whatever.

**Comment by jbarber2016** [comment]
are you using angular cli?

**Comment by damonus-prime** [comment]
No.  The project is generated by JHipster.

Not sure if it uses the cli under the hood, tho.  This is my first JHipster project, and it was already generated when I started working on it.

**Comment by jbarber2016** [comment]
I'm not too familiar with JHipster.
use angular cli:  it's far bettter.  
1) npm install -g @angular-cli
2) npm install --save-dev @angular-cli

now from command prompt
1) ng new TestProj --styles=css --routing
2) npm install --save ag-grid
3) npm install --save ag-grid-angular
4) ng g c components/test
5) put your html for <ag-grid-angular>
6) put your typescript
7) ng serve --open to launch locally

**Comment by jbarber2016** [comment]
also look at https://github.com/angular/angular-cli for all the different commands you can do.

**Comment by damonus-prime** [comment]
Super familiar with angular cli, as I use it for all my personal projects.  Sadly, this one is for work and I don't have that option.  :(

**Comment by makinggoodsoftware** [comment]
This ticket is flagged as managed_by_the_comunnity, this means that ag-grid staff is not going to be actively looking into it, and it will get closed if inactive for more than one month, the community is welcome to help with this question/support issue

===============

If you are an ag-Grid Enterprise Customer, please raise this in our Zendesk support system - contact info@ag-grid.com for access. This is our primary channel for Support.

Bugs Raised Here:

We will do a periodic review of GitHub issues for any bugs raised and these can be tracked on our pipeline page: https://www.ag-grid.com/ag-grid-pipeline

Feature Requests/Support Raised Here:

We are reviewing how we process Feature Requests that come from GitHub. We currently have a very full pipeline of work which will be our priority for the coming months. 

As a result, we don’t have capacity to take feature requests from GitHub.

We will review this as we move through our current backlog.

Thanks for your understanding.
ag-Grid Team.

**Comment by amazingkapil** [comment]
This is a frustrating error  I am facing now. @ShankzPatil  Sir can you tell me how do you attach scope.watch.

**Comment by ShankzPatil** [comment]
please show code snippet (plunker, codepen if possible), so i will able to solve the exact problem.

**Comment by amazingkapil** [comment]
`export class ViewComponent implements OnInit {
   
   private gridOptions: GridOptions;
   public rowData: client[];
   private columnDefs: any[];
    constructor(private cs: AllClientService){

        this.gridOptions = <GridOptions>{};
      
        this.columnDefs = this.createColumnDefs();
        this.rowData = this.createRowData();
        this.gridOptions={
            rowData: this.rowData,
            columnDefs: this.columnDefs,
            animateRows: true,
            rowSelection: 'single',
            enableColResize: true,
            floatingFilter: true,
            context: { componentParent: this },
             onGridReady: function(params) {
                 console.log("grid is started");  
     
             },
            overlayLoadingTemplate : "Loading..",
            overlayNoRowsTemplate : "No Records found"
        }
    }
    
    ngOnInit(){
        console.log("I before calling the service");
        console.log("I am inside the ngOnInit and loading the client");
        this.loadClients();
        
    }
    public loadClients():void{
        console.log("loading startedddd");
        this.gridOptions.api.setRowData(this.createRowData());   ### __ #  **//Error comes here ..it says api not defined**
        this.gridOptions.api.hideOverlay();
       }

    createRowData() {
        var rowData: client[]=[];
                rowData.push({
                    firstName:"Akash",
                    lastName:"Kapil",
                    email:"ash1998su@gmail.com",
                    SSN:"123-345-4567",
                    phone:"8302271549",
                    birthDay:"21-07-1994"
                });
        return rowData;
    }

    createColumnDefs() {
        return [
            {
                headerName: 'FirstName', width: 120, field: "firstName",  suppressFilter: true
            },
            {
                headerName: 'LastName', width: 120, field: "lastName", 
            },
            {
                headerName: 'Email', width: 57, field: "email",  suppressFilter: true
            },
            {
                headerName: 'SSN', width: 60, field: "SSN", suppressFilter: true
            },
            {
                headerName: 'phone', width: 100, field: "phone"
            },
            {
                headerName: 'birthday', width: 180, field: "birthDay"
            }
        ];
    }

}`

**Comment by ShankzPatil** [comment]
@amazingkapil find this plunker may helpful to you 
https://plnkr.co/edit/IZhsiNY2JB5DJ0uCc5lG?p=preview

**Comment by luisdelae** [comment]
Not sure if you've fixed this already or not, but I was having the same issue just now. I had to not only initialize gridOptions as an empty <GridOptions> but also assign a value to any of its properties within your constructor: 

```
this.gridOptions = <GridOptions>{ };
this.gridOptions.columnDefs = [{headerName: 'Test', firled: 'test'}]
```

**Comment by oripka** [comment]
I had a similar problem using Vue as a framework. In my case I changed the row data based on an incoming event via the Vue eventHub. I forgot to remove the event for the destroyed grid from the eventHub. Everything worked fine, because on the component (re-)creation I signed up for the events again. But the old destroyed component still got the event as well and caused errors.

In the end I just had to do:

beforeDestroy() {
    this.$eventHub.$off('updateGrid');
},
